### PR TITLE
Adding a feature searching the latest catalog

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward giantswarm/giantswarm#6824

it's copy from https://github.com/giantswarm/e2esetup/pull/40

### CODE REVIEW

File: catalog/catalog.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+func getIndex(catalog string) (index, error) {
+	indexURL := fmt.Sprintf("https://giantswarm.github.io/%s/index.yaml", catalog)
+	resp, err := http.Get(indexURL)
+	if err != nil {
+		return index{}, microerror.Mask(err)
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (tomahawk28) [inline catalog/catalog.go]: I mean to make it separate commit for these files but failed it 🤦‍♂

**Comment by TheoBrigitte** [comment]
I think using go modules here would be better.

**Comment by tomahawk28** [comment]
@TheoBrigitte Good idea, I will move it to go module in here.

**Comment by tomahawk28** [comment]
it works! 🎉

File: appcatalog.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/giantswarm/microerror"
+)
+
+type index struct {
```
Comment (rossf7) [inline appcatalog.go]: Let's move these to types.go to keep things tidy.

File: appcatalog.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	index, err := getIndex(catalog)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	var version string
+	{
+		entry, ok := index.Entries[app]
+		if !ok {
+			return "", microerror.Maskf(notFoundError, fmt.Sprintf("no app %q in index.yaml", app))
```
Comment (rossf7) [inline appcatalog.go]:
```suggestion
			return "", microerror.Maskf(notFoundError, fmt.Sprintf("no app %#q in index.yaml", app))
```

File: appcatalog.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		}
+		version = entry[0].Version
+	}
+
+	return version, nil
+}
+
+func getIndex(catalog string) (index, error) {
+	indexURL := fmt.Sprintf("https://giantswarm.github.io/%s/index.yaml", catalog)
```
Comment (rossf7) [inline appcatalog.go]: I'd like to remove the hardcoding if we can. The appcatalog CR already has the storage URL. So maybe we can pass that in and just append `index.yaml` here. WDYT?


```
k get appcatalog giantswarm-incubator -o yaml

apiVersion: application.giantswarm.io/v1alpha1
kind: AppCatalog
metadata:
  name: giantswarm-incubator
spec:
...
  storage:
    URL: https://giantswarm.github.com/giantswarm-incubator-catalog/
    type: helm
  title: Giant Swarm Incubator

File: appcatalog.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/giantswarm/microerror"
+)
+
+type index struct {
+	APIVersion string             `yaml:"apiVersion"`
+	Entries    map[string][]entry `yaml:"entries"`
+	Generated  time.Time          `yaml:"generated"`
```
Comment (kopiczko) [inline appcatalog.go]: With `ghodss/yaml` we rather need json tags.

Comment (kopiczko) [inline appcatalog.go]: Also why there are more fields than needed? I see only `Entries` being used.

File: appcatalog.go (hunk 4 above)
Comment (tomahawk28) [reply appcatalog.go]: it makes sense, I will change the parameter name as `storage`.

File: appcatalog.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	index, err := getIndex(catalog)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	var version string
+	{
+		entry, ok := index.Entries[app]
+		if !ok {
+			return "", microerror.Maskf(notFoundError, fmt.Sprintf("no app %#q in index.yaml", app))
```
Comment (kopiczko) [inline appcatalog.go]:
```suggestion
			return "", microerror.Maskf(notFoundError, "no app %#q in index.yaml", app)
```

File: appcatalog.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	index, err := getIndex(storage)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	var downloadURL string
+	{
+		entry, ok := index.Entries[app]
+		if !ok {
+			return "", microerror.Maskf(notFoundError, fmt.Sprintf("no app %q in index.yaml", app))
```
Comment (kopiczko) [inline appcatalog.go]:
```suggestion
			return "", microerror.Maskf(notFoundError, "no app %#q in index.yaml", app)
```

File: appcatalog.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ghodss/yaml"
+	"github.com/giantswarm/microerror"
+)
+
+// GetLatestChart returns the latest chart tarball file in the specified catalog.
+func GetLatestChart(ctx context.Context, storage, app string) (string, error) {
```
Comment (kopiczko) [inline appcatalog.go]: Is storage here an URL? if so maybe `storageURL` because it isn't clear what it is. Also maybe describing that in godoc would be helpful.

Comment (kopiczko) [inline appcatalog.go]: I also would like to have described when this returns notFoundError in the godoc. Also applies to the other public function.

Comment (tomahawk28) [reply appcatalog.go]: Sure, I just added new godoc but probably need someone's help for fixing 🇺🇸nit.

File: appcatalog.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ghodss/yaml"
+	"github.com/giantswarm/microerror"
+)
+
+// GetLatestChart returns the latest chart tarball file in the AppCatalog storage URL
```
Comment (rossf7) [inline appcatalog.go]:
```suggestion
// GetLatestChart returns the latest chart tarball file for the specified storage URL and app
```

This is good just suggesting a minor change.

File: appcatalog.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+		if !ok {
+			return "", microerror.Maskf(notFoundError, "no app %#q in index.yaml", app)
+		}
+		downloadURL = entry[0].Urls[0]
+	}
+
+	return downloadURL, nil
+}
+
+// GetLatestVersion returns the latest app version in the AppCatalog storage URL
```
Comment (rossf7) [inline appcatalog.go]:
```suggestion
// GetLatestVersion returns the latest app version for the specified storage URL and app
```

File: go.mod [hunk 11]
```diff
@@ -0,0 +1,10 @@
+module github.com/giantswarm/appcatalog
+
+go 1.12
```
Comment (rossf7) [inline go.mod]: I think this should be aligned with devctl. So the go version is last and is 1.13.

https://github.com/giantswarm/devctl/blob/f39ef486eb7dd167213e2e705e0c25539a4ef08a/go.mod

Also... I need to upgrade my go version. 😅

File: appcatalog.go (hunk 9 above)
Comment (tomahawk28) [reply appcatalog.go]: Nice, Thanks for noticing.

**Comment by tomahawk28** [comment]
I will follow with a next PR to make `NewTarballURL` function.
